pkg/utils/shared_objects/host_symbols_loader.go

```diff
@@ -0,0 +1,131 @@
+package shared_objects
+
+import (
+	"debug/elf"
+	"github.com/hashicorp/golang-lru/simplelru"
+)
+
+// HostSharedObjectSymbolsLoader is responsible for efficient reading of shared object's symbols.
+// The logic of the loader here is used on absolute paths, so container relative paths won't work here.
+type HostSharedObjectSymbolsLoader struct {
+	loadingFunc func(path string) (*soDynamicSymbols, error)
+	soCache     soDynamicSymbolsCache
+}
+
+func InitSODynamicSymbolsLoader(cacheSize int) *HostSharedObjectSymbolsLoader {
+	lruCallback := simplelru.EvictCallback(func(key interface{}, value interface{}) {})
+	sharedObjectsLRU, _ := simplelru.NewLRU(cacheSize, lruCallback)
+	soCache := dynamicSymbolsLRUCache{sharedObjectsLRU}
+	return &HostSharedObjectSymbolsLoader{
+		soCache:     &soCache,
+		loadingFunc: loadSharedObjectDynamicSymbols,
+	}
+}
+
+// GetDynamicSymbols try to get shared objects dynamic symbols from lru, and if fails read needed information
+// from ELF file.
+func (soLoader *HostSharedObjectSymbolsLoader) GetDynamicSymbols(soInfo SoExaminationInfo) (map[string]bool, error) {
+	syms, err := soLoader.loadSOSymbols(soInfo)
+	if err != nil {
+		return nil, err
+	}
+	dynSyms := copyMap(syms.ImportedSymbols)
+	for expSym := range syms.ExportedSymbols {
+		dynSyms[expSym] = true
+	}
+	return dynSyms, nil
+}
+
+// GetExportedSymbols try to get shared objects exported symbols from lru, and if fails read needed information
+// from ELF file.
+func (soLoader *HostSharedObjectSymbolsLoader) GetExportedSymbols(soInfo SoExaminationInfo) (map[string]bool, error) {
+	syms, err := soLoader.loadSOSymbols(soInfo)
+	if err != nil {
+		return nil, err
+	}
+	return copyMap(syms.ExportedSymbols), nil
+}
+
+// GetImportedSymbols try to get shared objects imported symbols from lru, and if fails read needed information
+// from ELF file.
+func (soLoader *HostSharedObjectSymbolsLoader) GetImportedSymbols(soInfo SoExaminationInfo) (map[string]bool, error) {
```

Comment:
To bring another perspective to this: I very much like immutability, it reduces surface for misuse of code and bugs later down the line. 
But we could be in a performance sensitive context here. If these functions will used in hot paths of the code, we should probably avoid the copy, but that should be benchmarked later on, when we know how this code will be used. Otherwise I think it's a premature optimization.